Add async scan algorithms

allisonvacanti opened the pull request:
Prerequisite: NVIDIA/cub#210

allisonvacanti commented on testing/async_exclusive_scan.cu:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  // Invoke the async algorithm for a single overload as described by
+  // the prefix and postfix tuples. These tuple contains instances of any
```
s/These tuple/These tuples/

allisonvacanti commented:
The "large indices" tests are currently failing because temporary memory allocation fails for the larger problem sizes. This doesn't happen for the synchronous scan (which uses a custom implementation instead of just calling `cub::DeviceScan`). I'll pull those tests out before merging if I can't find a solution/workaround.

trivialfis commented on thrust/system/cuda/detail/async/exclusive_scan.h:
```diff
@@ -0,0 +1,180 @@
+
+  // Run scan.
+  thrust::cuda_cub::throw_on_error(
+    cub::DeviceScan::ExclusiveScan(
```
Sorry for the many questions.  From this https://github.com/NVIDIA/thrust/issues/1299#issuecomment-704024397, should we explicitly specify the size type via cub dispatching?

allisonvacanti replied:
Yep, that seems to be the issue with the failing test. I'll need to update this to use those APIs instead.

allisonvacanti commented:
This is ready for review, but will need NVIDIA/cub#213 merged first so we can test without hitting timeouts.

brycelelbach commented on testing/async/exclusive_scan/basic.cu:
```diff
@@ -0,0 +1,72 @@
+};
+
+template <typename T>
+struct TestBasic
```
We really ought to unify naming conventions to `snake_case`; I prefer that new stuff at least use `snake_case`

brycelelbach replied:
I'd prefer to have `exclusive_scan` in the name to be consistent with other tests.

brycelelbach replied:
And perhaps call this "simple", because I think that's the convention of the other tests (instead of "basic").

allisonvacanti replied:
1. Changed to `snake_case` for the all new async test ✔
2. `exclusive_scan` appears in the executable and test name. These are subtests of `exclusive_scan`, including the algorithm name for each subtest here would be redundant and verbose.
3. Changed `basic` to `simple`.

brycelelbach commented on testing/async/exclusive_scan/large_indices.cu:
```diff
@@ -0,0 +1,239 @@
+
+// Fake iterator that asserts
+// (a) it is written with a sequence and
+// (b) a defined maximum value is written at some point
```
Can we leave a note here explaining this is necessary to test that larger than 32bit indices work without actually needing storage for that many indices (which would require a lot of GPU memory).

allisonvacanti replied:
Done.

brycelelbach commented on testing/async/exclusive_scan/large_indices.cu:
```diff
@@ -0,0 +1,239 @@
+  mutable thrust::device_ptr<bool> unexpected_value{nullptr};
+
+  // Should be called on the first iterator generated.
+  void initialize_shared_state()
```
Why is not a constructor? Leave a note explaining.

brycelelbach replied:
@griwes

allisonvacanti replied:
Added a note.

brycelelbach commented on testing/async/exclusive_scan/large_indices.cu:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  // Should be called only once on the initialized iterator.
+  void free_shared_state() const
```
Why is not a destructor? Leave a note explaining.

brycelelbach replied:
@griwes

allisonvacanti replied:
Added a note.

brycelelbach commented on testing/async/exclusive_scan/large_indices.cu:
```diff
@@ -0,0 +1,239 @@
+// Fake iterator that asserts
+// (a) it is written with a sequence and
+// (b) a defined maximum value is written at some point
+struct assert_sequence_iterator
```
I'd prefer to have a single implementation of this for both algorithms if possible to reduce code duplication.

allisonvacanti replied:
These iterator implementations are unique enough that we need different ones for the different scans.

I may try to abstract this later as we expand the new-style tests to more algorithms, we'll see how that goes.

brycelelbach commented on testing/async/exclusive_scan/large_types.cu:
```diff
@@ -0,0 +1,63 @@
+
+    test_policy_overloads<invoker<FixedVector<int, 1>>>::run(num_values);
+
+#ifndef __QNX__ // These are excluded on QNX for scan.cu's version of this test.
```
I'm not sure why this was failing only on QNX, but I think we should remove this ifdef and see what happens!

allisonvacanti replied:
Done.

brycelelbach commented on testing/async/exclusive_scan/mixin.h:
```diff
@@ -0,0 +1,116 @@
+} // namespace invoke_reference
+
+//------------------------------------------------------------------------------
+namespace invoke_async
```
Can you give me an overview of the namespace/naming scheme used here?

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+//
+//    //-------------------------------------------------------------------------
+//    //
+//    testing::async::mixin::exclusive_scan::mixin::postfix_args::all_overloads
```
Can the second `mixin` be dropped here?

allisonvacanti replied:
Turns out this was just a typo in the docs -- the actual implementation only uses a single `mixin` in the namespace path.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+                                     validate,
+                                     num_values);
+    }
+    catch (unittest::UnitTestException &exc)
```
We have a bit of a bug when catching exceptions from non-CUDA backends, notably TBB and OpenMP; they wrap exceptions that occur in their threads in a special exception type (`tbb::captured_exception`, etc) which means these catch blocks won't match.

allisonvacanti replied:
The async framework is CUDA-only for the foreseeable future. If it gets expanded to other backends we can address this at that time.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+                               output_ref,
+                               postfix_tuple,
+                               postfix_index_seq{});
+
```
We should probably leave a comment here explaining that waiting happens in `compare_outputs`

allisonvacanti replied:
Done.

brycelelbach commented on thrust/system/cuda/detail/async/scan.h:
```diff
@@ -0,0 +1,33 @@
+/******************************************************************************
```
Do we really need this header? Can we get rid of it and just use granular includes?

allisonvacanti replied:
This matches the existing convention used for the sync scans and simplifies the ADL dispatch headers a bit. Let's consider the "separate algorithms per header" as an implementation detail of the CUDA backend here. I think it'd be nice to eventually split the headers out for both sync/async scans at some point as a cleanup.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+//          all_overloads<initial_value_type, alternate_binary_op>
+//      , testing::async::exclusive_scan::mixin::invoke_reference::
+//          host_synchronous<input_value_type, output_value_type>
+//      , testing::async::exclusive_scan::mixin::invoke_async::basic
```
We should call `basic` simple because `simple` is the name used in existing.

allisonvacanti replied:
Done.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+//
+//    //-------------------------------------------------------------------------
+//    // testing::async::mixin::input::device_vector
+//    using input_type  = thrust::device_vector<input_value_type>;
```
There's an extra space here: `input_type  =`.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+//      std::tuple<>,                                       // - no extra args
+//      std::tuple<initial_value_type>,                     // - initial_value
+//      std::tuple<initial_value_type, alternate_binary_op> // - initial_value,
+//      binary_op
```
`binary_op` appears to have been line wrapped, it should be on the previous line.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+{
+
+// TODO it'd be nice to specify a lambda expression that'd replace the call to
+// thrust::sequence when desired.
```
The idea here is to parameterize initialization?

allisonvacanti replied:
Yep, it'd be a nice feature to have at some point, but there are easy workaround for now.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+};
+
+template <typename value_type>
+struct counting_iterator_0
```
Could this be `_starts_at_0` or `_from_0`, etc?

allisonvacanti replied:
Changed to `_from_0`.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+  {
+    using iterator = thrust::discard_iterator<>;
+
+    iterator begin() { return thrust::make_discard_iterator(); }
```
I don't think we need the non-const here.

allisonvacanti replied:
Removed.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+  }
+};
+
+struct discard_iterator
```
@griwes suggests calling these `*_range`.

allisonvacanti replied:
The current name matches the name of what we're testing. I prefer this as-is.

It's scoped to a namespace so we can always differentiate this from the thrust class.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(1)}; }
```
I don't think we need the non-const here.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+
+    std::size_t num_values;
+
+    iterator begin() { return iterator{static_cast<value_type>(1)}; }
```
I don't think we need the non-const here.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+    iterator begin() const { return iterator{static_cast<value_type>(1)}; }
+    iterator cbegin() const { return iterator{static_cast<value_type>(1)}; }
+
+    iterator end() { return iterator{static_cast<value_type>(1 + num_values)}; }
```
I don't think we need the non-const here.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+template <typename value_type>
+struct constant_iterator_1
+{
+  struct input_type
```
We should probably explain somewhere that the `input_type`s need random access (thus begin, end, and a size), while the `output_type`s (like discard) only need to satisfy output iterator / range.

allisonvacanti replied:
Added this info to the appropriate sections in the class documentation.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+ * // algorithm([policy,] input, output, initial_value)
+ * // algorithm([policy,] input, output, initial_value, binary_op)
+ * template <typename value_type, typename alternate_binary_op =
+ * thrust::maximum<>> struct all_overloads
```
Bad line wrapping.

allisonvacanti replied:
Fixed.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+ * thrust::maximum<>> struct all_overloads
+ * {
+ *   using postfix_args_type = std::tuple<         // List any extra arg
+ * overloads: std::tuple<>,                               // - no extra args
```
Bad line wrapping.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+
+void basic_event_validation(thrust::device_event& e)
+{
+  TEST_EVENT_WAIT(e);
```
Do we still need the macros if we're just abstracting this into a function? Can we nuke the macros?

allisonvacanti replied:
That sounds like a great candidate for a follow-up PR.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+
+// Does an 'almost_equal' comparison for floating point types, since fp
+// addition is non-associative
+struct assert_maybe_fuzzy_equal
```
Can we call this `assert_almost_equal`?

allisonvacanti replied:
Updated to `assert_almost_equal_if_fp`.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+                                   std::false_type /* is_floating_point */)
+  {
+    detail::basic_event_validation(e);
+    ASSERT_EQUAL(ref, test);
```
Can we use `ASSERT_ALMOST_EQUAL` for integer types?

allisonvacanti replied:
I avoid that because it's not necessary here -- these are for the scan algorithms which give exact results for integer types but approximate results for fp due to non-associative addition. So checking with exactly equal for integers is intentional here, see the comment on the struct containing this implementation.

brycelelbach commented on testing/async/mixin.h:
```diff
@@ -0,0 +1,645 @@
+
+// Does an 'almost_equal' comparison for floating point types, since fp
+// addition is non-associative
+struct assert_maybe_fuzzy_equal_quiet
```
s/maybe_fuzzy/almost/

allisonvacanti replied:
Updated as above.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+  static void run_basic_policy_test(std::string const &policy_desc,
+                                    PolicyTuple &&policy_tuple,
+                                    ValidateEvent &&validate,
+                                    std::size_t const num_values)
```
I have slight preference for this being combined with the other `run_basic_policy_test` overload, or call one of them `_impl`; they don't do the same thing, so they should not be overloads.

allisonvacanti replied:
These aren't overloads -- `run_basic_policy_tests` (plural) launches tests for each policy configurations by calling `run_basic_policy_test` (singular).

But I do like your idea of doing a function-scope try block in `basic_policy_test` instead of having the `run_basic_policy_test` trampoline, so I'll do that.

brycelelbach commented:
Few high-level remarks:

* I'd like this to support input initialization on the host because there are some initialization strategies that are host only, like `unittest::random_integer`.
* I'd prefer to only generate input once and then copy.
* I'd like to retain the `unittest::random_integer` tests for async algorithms.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+  //----------------------------------------------------------------------------
+  // For the specified postfix overload set, test the algorithm with several
+  // different policy configurations.
+  template <std::size_t postfix_idx>
```
Template parameters should be `CamelCase`.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+  //----------------------------------------------------------------------------
+  // Test .after(event/future) handling:
+  template <std::size_t postfix_idx>
+  static void run_after_future_test(std::size_t const num_values)
```
Let's try to unify the naming scheme. We have:
* Two overloads for `run_basic_policy_test`, one which has the try block, one which doesn't. Then there's also `basic_policy_test`.
* `run_after_future_test` and `after_future_test`, which appear to have the same relationship as the `run_basic_policy_test` overloads.

allisonvacanti replied:
Done.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+  // For the specified postfix overload set, test the algorithm with several
+  // different policy configurations.
+  template <std::size_t postfix_idx>
+  static void run_basic_policy_tests(std::size_t const num_values)
```
Let's just call this one `basic_policy_test` to mirror `after_future_test`.

allisonvacanti replied:
Renamed so that both tests have a `run_XXX_tests` entry point, which can launch multiple `XXX_test` implementations when needed.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+    std::tuple_element_t<postfix_idx, postfix_args_type>;
+    postfix_tuple_type const postfix_tuple = get_postfix_tuple<postfix_idx>();
+
+    // Generate index sequences for the tuples. Prefix size always = 1 here.
```
Can we expand on this comment here explaining why prefix size is always 1?

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+                        thrust::event_error(thrust::event_errc::no_state));
+
+    // Explicitly order this invocation after e_b:
+    auto policy_after_e_b = thrust::device.after(e_b);
```
We should leave a comment here explaining that it is important that in the first of these tests, the execution policy is an rvalue and in the second of these tests it is an lvalue; both are supposed to be consumable.

allisonvacanti replied:
Done.

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+
+    // Explicitly order this invocation after e_b:
+    auto policy_after_e_b = thrust::device.after(e_b);
+    // Make sure the tuple contains an lvalue ref:
```
Can we say something like "we can't use make_tuple because it would make a copy".

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+    // Should use same stream as e_b:
+    ASSERT_EQUAL_QUIET(stream_b, stream_c);
+
+    // Verify that indirect double consumption of e_b produces an exception:
```
s/indirect/lvalue/

brycelelbach commented on testing/async/test_policy_overloads.h:
```diff
@@ -0,0 +1,421 @@
+    // Second invocation should use same stream as before:
+    ASSERT_EQUAL_QUIET(stream_a, stream_b);
+
+    // Verify that double consumption of e_a produces an exception:
```
Say "rvalue double consumption" to be consistent with below.

allisonvacanti replied:
Clarified value categories in comments.

brycelelbach commented on thrust/system/cuda/detail/async/exclusive_scan.h:
```diff
@@ -0,0 +1,199 @@
+/******************************************************************************
+ * Copyright (c) 2016, NVIDIA CORPORATION.  All rights reserved.
```
Adjust copyright dates.

brycelelbach commented on thrust/system/cuda/detail/async/exclusive_scan.h:
```diff
@@ -0,0 +1,199 @@
+                            OutputIt out,
+                            InitialValueType init,
+                            BinaryOp op)
+  -> unique_eager_event
```
Don't use a trailing return type that doesn't depend on anything.

brycelelbach commented on thrust/system/cuda/detail/async/exclusive_scan.h:
```diff
@@ -0,0 +1,199 @@
+                                       OutputIt,
+                                       BinaryOp,
+                                       InitialValueType,
+                                       thrust::detail::int64_t>;
```
Ouch, I don't think `async::reduce` does dispatch correctly; can we either do this in this patch, or file an issue to track that?

allisonvacanti replied:
This will be fixed with NVIDIA/cub#212.

brycelelbach commented on thrust/system/cuda/detail/async/inclusive_scan.h:
```diff
@@ -0,0 +1,195 @@
+                                       BinaryOp,
+                                       cub::NullType,
+                                       thrust::detail::int32_t>;
+
```
There was no newline between these two `usings` in `exclusive_scan`.

brycelelbach commented on thrust/async/scan.h:
```diff
@@ -0,0 +1,355 @@
+/*
```
We should have separate headers for each of the algorithms.

allisonvacanti replied:
I'm going to hold off on this -- we should think about more granular headers in general in thrust. Since the `thrust::` scan algorithms are all in a single public header, I'd rather the async public headers match those conventions.

brycelelbach commented on thrust/async/scan.h:
```diff
@@ -0,0 +1,355 @@
+             Sentinel&& last,
+             OutputIt&& out,
+             BinaryOp&& op) const
+    // ADL dispatch.
```
Extra spaces before `//`.

brycelelbach commented on testing/async/inclusive_scan/using_vs_adl.cu:
```diff
@@ -0,0 +1,169 @@
+      inclusive_scan(host_input.cbegin(),
+                     host_input.cend(),
+                     host_output.begin(),
+                     std::get<PostfixArgIndices>(THRUST_FWD(postfix_tuple))...);
```
I don't think synchronous scans are marked nodiscard, so I think we can just ignore the result altogther.

allisonvacanti replied:
Good point, I got carried away following the reduce algorithms as a reference :) I've removed the nodiscard annotations completely.

brycelelbach commented on testing/async/inclusive_scan/large_types.cu:
```diff
@@ -0,0 +1,63 @@
+
+    test_policy_overloads<invoker<FixedVector<int, 1>>>::run(num_values);
+
+#ifndef __QNX__ // These are excluded on QNX for scan.cu's version of this test.
```
Take this out, it should no longer be a problem on QNX. Maybe file an issue to take this out from the synchronous scan tests too?

allisonvacanti replied:
#1345

brycelelbach commented on testing/async/inclusive_scan/mixed_types.cu:
```diff
@@ -0,0 +1,95 @@
+{
+  // 10 values are enough to check the behavior we want to test while staying
+  // small enough to reason about during debugging.
+  constexpr std::size_t num_values = 10;
```
This is incorrect, because CUB has different codepaths for single tile versus multi-tile; we need to test both, so this should use the testing mechanism that supports a variable problem size (ditto for the corresponding exclusive_scan test).

allisonvacanti replied:
I added a `DECLARE_SIZED_UNITTEST` macro and used it for these. The new macro was needed since the others all expect the test to take template parameters.

hcedwar commented:
The "files changed" code volume is quite large.  Where is an algorithm & design summary to review this aspect vs. code details?

allisonvacanti commented:
They're mostly changes to Thrust's internal testing infrastructure. The actual API changes here are very small and follow the existing conventions and design plan for Thrust's existing async algorithms.

There's a design doc from a couple years ago floating around for the async design. If you need to see it for some reason I can dig it up.

allisonvacanti commented:
We discussed this in a call, but to update the PR review:

> * I'd like this to support input initialization on the host because there are some initialization strategies that are host only, like `unittest::random_integer`.

The input can be initialized on the host and copied to the device if a host-only strategy is needed. Since the input is more frequently accessed on the device (4 device invocations vs 1 host invocation), this reduces the amount of transfer overhead in the common case that inputs can be generated on the device.

> * I'd prefer to only generate input once and then copy.

Me too, but I'd also like to not require copyability for the generic `input_type` concept. If this ends up being problematic in practice I'm happy to revisit this.

> * I'd like to retain the `unittest::random_integer` tests for async algorithms.

Can do. I'll see about adapting it to run on the device, too -- we have device-side rng facilities, so hopefully it'll be trivial.

allisonvacanti commented:
Addressed feedback. The `mixed_types` tests are failing now, I'll need to investigate this. We also discussed shortening the namespace for some of the device-specific test mixins, I'll fix that when I take a look at the new test failures.

allisonvacanti commented:
`mixed_types` test should still be failing, but running DVS to try to repro build timeouts.

DVS CL: 29371528

allisonvacanti commented:
Latest round of DVS runs show the following issues on Clang:

```
thrust/system/cuda/detail/async/exclusive_scan.h:160:8: error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]
thrust/system/cuda/detail/async/inclusive_scan.h:158:8: error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]
```

allisonvacanti commented:
run tests

brycelelbach commented:
run tests

allisonvacanti commented:
run tests

allisonvacanti commented:
run tests

allisonvacanti commented:
DVS CL: 29473661
